app-ng/src/handlers.lisp

```diff
@@ -0,0 +1,179 @@
+          :do (return-from dispatch-rpc-handlers handler)))
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun remove-existing-rpc-handlers (rpc-method name)
+    (remove-if (lambda (handler-spec)
+                 (or (eq name (cdr handler-spec))
+                     (string= rpc-method (car handler-spec))))
+               *rpc-handlers*))
+
+  (defun make-json-parameter (var)
+    `(,var (and (boundp '*request-json*) (json-parameter ,(string-downcase (symbol-name var))))))
+
+  (defun make-parsed-binding (parameter-spec)
+    (destructuring-bind (var parse-function) parameter-spec
+      `(,var (funcall (alexandria:ensure-function ,parse-function) ,var))))
+
+  (defun valid-rpc-handler-lambda-list-p (lambda-list)
+    (every (lambda (parameter-spec)
+             (and (= 2 (length parameter-spec))
+                  (symbolp (first parameter-spec))))
+           lambda-list)))
+
+(defmacro define-rpc-handler ((handler-name rpc-method) lambda-list &body body)
+  "Define an RPC handler named HANDLER-NAME for the RPC method RPC-METHOD.
+
+DEFINE-RPC-HANDLER is like TBNL:DEFINE-EASY-HANDLER, but modified to dispatch on the RPC-METHOD name rather than URI path and to default parameters in LAMBDA-LIST based on JSON parameters rather than GET/POST parameters.
+
+HANDLER-NAME is a SYMBOL and becomes the name of the FUNCTION that gets defined.
+
+RPC-METHOD is a STRING that defines the RPC method name that is used to call this handler. DISPATCH-RPC-HANDLERS will check the caller-provided JSON request body and dispatch to this handler if the \"type\" field is STRING= to RPC-METHOD.
+
+LAMBDA-LIST is a LIST of (VAR-NAME PARSE-FUNCTION) pairs. Each VAR-NAME will expand to an &KEY parameter in the lambda-list of HANDLER-NAME, and will be bound to the caller-provided JSON request parameters with the corresponding name. The JSON parameter name is derived by converting VAR-NAME to a lowercase string. So, e.g., if VAR-NAME is FOO-BAR, then FOO-BAR is bound in BODY to the value of (JSON-PARAMETER \"foo-bar\"). This defaulting of VAR-NAME only occurs in contexts where *REQUEST-JSON* is bound; otherwise, the value of VAR-NAME defaults to NIL. Parameter defaulting is done in this way to allow HANDLER-NAME to be conveniently called outside the server context with the caller providing the values of the keyword arguments directly.
+
+Additionally, the specified PARSE-FUNCTION will be called to parse and validate the caller-provided value. PARSE-FUNCTION should raise an ERROR for invalid inputs, and should return a parsed value of whatever type is required in the BODY of the function. Note that PARSE-FUNCTION is called both for defaulted and non-defaulted arguments. The intention is to reduce argument unpacking/validation boilerplate in BODY and to make it difficult to forget to validate user input.
+
+EXAMPLE:
+
+    (define-rpc-handler (handle-frobnicate \"frobnicate\") ((number-of-times parse-non-negative))
```

Comment:
this is @rigettizach 's favorite function!

Reply:
Impeccable taste